atc/exec/retry_error_step.go

```diff
@@ -48,18 +50,15 @@ func (step RetryErrorStep) Run(ctx context.Context, state RunState) error {
 }
 
 func (step RetryErrorStep) toRetry(logger lager.Logger, err error) bool {
-	switch err.(type) {
-	case transport.WorkerMissingError, transport.WorkerUnreachableError:
+	var urlError *url.Error
+	if errors.As(err, &transport.WorkerMissingError{}) || errors.As(err, &transport.WorkerUnreachableError{}) || errors.As(err, &urlError) {
 		logger.Debug("retry-error",
 			lager.Data{"err_type": reflect.TypeOf(err).String(), "err": err.Error()})
 		return true
-	default:
-		re := regexp.MustCompile(`worker .+ disappeared`)
-		if re.MatchString(err.Error()) {
-			logger.Debug("retry-error",
-				lager.Data{"err_type": reflect.TypeOf(err).String(), "err": err})
-			return true
-		}
+	} else if _, ok := err.(net.Error); ok || regexp.MustCompile(`worker .+ disappeared`).MatchString(err.Error()) {
```

Comment:
You should use the same `errors.As` for `net.Error` so it works with nested steps - it works for interfaces, too!

i.e.

```go
var netError net.Error
if errors.As(err, &netError) {
    ...
}
```

See the following demo: https://play.golang.org/p/yHvlZ2V7ApQ (note that `err.(net.Error)` returns false since it's a wrapped error)

Reply:
Let me try